src/sentry/static/sentry/app/views/settings/components/forms/choiceMapper.jsx

```diff
@@ -0,0 +1,188 @@
+import {Flex, Box} from 'grid-emotion';
+import PropTypes from 'prop-types';
+import React from 'react';
+import styled from 'react-emotion';
+
+import {defined, objectIsEmpty} from 'app/utils';
+import {t} from 'app/locale';
+import Button from 'app/components/buttons/button';
+import DropdownAutoComplete from 'app/components/dropdownAutoComplete';
+import DropdownButton from 'app/components/dropdownButton';
+import InputField from 'app/views/settings/components/forms/inputField';
+import SelectControl from 'app/components/forms/selectControl';
+
+export default class ChoiceMapper extends React.Component {
+  static propTypes = {
+    ...InputField.propTypes,
+    /**
+     * Text used for the 'add row' button.
+     */
+    addButtonText: PropTypes.node,
+    /**
+     * Configuration for the add item dropdown.
+     */
+    addDropdown: PropTypes.shape(DropdownAutoComplete.propTypes).isRequired,
+    /**
+     * The label to show above the row name selected from the dropdown.
+     */
+    mappedColumnLabel: PropTypes.node,
+    /**
+     * A list of column labels (headers) for the multichoice table. This should
+     * have the same number of items as the mappedSelectors prop.
+     */
+    columnLabels: PropTypes.objectOf(PropTypes.node).isRequired,
+    /**
+     * A list of select field properties that should be used to render the
+     * select field for each column in the row.
+     */
+    mappedSelectors: PropTypes.objectOf(PropTypes.objectOf(SelectControl.propTypes))
+      .isRequired,
+  };
+
+  hasValue = value => defined(value) && !objectIsEmpty(value);
+
+  renderField = props => {
+    const {
+      onChange,
+      onBlur,
+      addButtonText,
+      addDropdown,
+      mappedColumnLabel,
+      columnLabels,
+      mappedSelectors,
+      disabled,
+    } = props;
+
+    const mappedKeys = Object.keys(mappedSelectors);
+    const emptyValue = mappedKeys.reduce((a, v) => ({...a, [v]: null}), {});
+
+    const valueIsEmpty = this.hasValue(props.value);
+    const value = valueIsEmpty ? props.value : {};
+
+    const addRow = data => {
+      onChange({...value, [data.value]: emptyValue}, {});
+      onBlur();
+    };
+
+    const removeRow = itemKey => {
+      const updatedValue = {...value};
```

Comment:
Since you're destructuring everywhere else, you can destructure with computed prop names:

```javascript
const {[itemKey]: _, ...updatedValue} = value;
```

`itemKey` gets destructured to alias `_`

Reply:
oh nice, didn't think about this

Reply:
a1a4184